cve_bin_tool/util.py

```diff
@@ -334,6 +335,122 @@ def validate_location(location: str) -> bool:
     return bool(re.match(pattern, location))
 
 
+def decode_purl(purl: str) -> ProductInfo | None:
+    """
+    Validate and decode a Package URL (purl) in the format: pkg:type/namespace/product@version.
+
+    Args:
+    - purl (str): Package URL (purl) string.
+
+    Returns:
+    - ProductInfo | None: An instance of ProductInfo containing the vendor, product,
+      version, location, and purl, or None if the purl is invalid.
+    """
+    # Define the regex pattern for the expected format
+    pattern = re.compile(r"^pkg:[^/]+/([^/]+)/([^@]+)@(.+)$")
```

Comment:
Couple of questions:
- does purl allow for escaping of / and @ at all?  (If it does, we'll need to change this regex)
- does the purl conform to the usual url specs?  
- could we be using urllib or similar to parse this?

Reply:
As a rule of thumb: 
- regexes for quick and dirty shell scripts you use to test a theory and will only use a few times
- well-tested libraries for anything you're going to support for months or years 

I love regexes (I learned them before I learned Python!) but they are a security/denial of service/maintainability nightmare, and we're probably using them too much in cve-bin-tool already (having just re-looked at the language parsers, I have some refactoring notes, but they're hardly the only spots).  Some of that's just "not many people supported this when we started."  But it sounds like we don't have to create and maintain them here, so yay!

Reply:
@terriko real thanks for the insight you shared on regex, I'll always remember it while working with regex, I have removed the regex matching can you take a look.